lib/rubygems/security.rb

```diff
@@ -338,27 +338,16 @@ module Gem::Security
 
   class Exception < Gem::Exception; end
 
-  ##
-  # Digest algorithm used to sign gems
-
-  DIGEST_ALGORITHM =
-    if defined?(OpenSSL::Digest::SHA256)
-      OpenSSL::Digest::SHA256
-    elsif defined?(OpenSSL::Digest::SHA1)
-      OpenSSL::Digest::SHA1
-    else
-      require 'digest'
-      Digest::SHA512
-    end
-
   ##
   # Used internally to select the signing digest from all computed digests
 
   DIGEST_NAME = # :nodoc:
-    if DIGEST_ALGORITHM.method_defined? :name
-      DIGEST_ALGORITHM.new.name
+    if defined?(OpenSSL::Digest::SHA256)
+      'SHA256'
+    elsif defined?(OpenSSL::Digest::SHA1)
+      'SHA1'
     else
-      DIGEST_ALGORITHM.name[/::([^:]+)\z/, 1]
+      'SHA512'
     end
```

Comment:
All code paths using Gem::Security::DIGEST_{ALGORITHM,NAME}, or Gem::Security in general, seem to have required OpenSSL anyway - Gem::Security::Signer#sign wouldn't work if DIGEST_ALGORITHM is ::Digest::SHA512.

Since `defined?(OpenSSL::Digest::SHA256)` is never false when OpenSSL is available, I think this can simply be

```
DIGEST_NAME = 'SHA256'
```

Reply:
I wasn't entirely sure of the original logic but this is the closest I could come up with while fixing the deprecations and keeping the behaviour intact.

I dug a little deeper in the history:
- [Allow RubyGems tests to run without OpenSSL (2013)](https://github.com/rubygems/rubygems/commit/0bb225a6063df0448a28f8bfe147422f0d7d1c24)
- [Gem::Security::DIGEST_ALGORITHM updated to SHA-256. (2016)](https://github.com/rubygems/rubygems/commit/d35a6c3b02f5b1d020c27b8ef2d412fdbcafde4b)
- [Fallback to Digest::SHA512 (2018)](https://github.com/rubygems/rubygems/commit/9471f8ed2bdc12248d2619bbbce6e53cd6c16cb6) - this one surprised me, why not `Digest::SHA256` instead?

I think your suggestion of `DIGEST_NAME = 'SHA256'` is safe to do since the `create_digest` method now takes care of situations where OpenSSL is unavailable. It also decouples algorithm choice from implementation availability - SHA256 is supported in both `Digest` and `OpenSSL::Digest`, and keeps the TarWriter specific fallback contained to that use case. I will add another commit with this refactor.